File: python/LinkedList/ReverseLinkedList.py
"""
206. Reverse Linked List
Difficulty: Easy
Topics: LinkedList, Recursion
"""
import ListNode

def reverseList(head: ListNode) -> ListNode:
    if not head:
        return head
    prev = None
    cur = head

    while cur:
        tmp = cur.next
        cur.next = prev
        prev = cur
        cur = tmp

    return prev

File: java/LinkedList/ReverseLinkedList.java
package java.LinkedList;

class ReverseLinkedList {

    /**
     * 206. Reverse Linked List
     * Difficulty: Easy
     * Topics: Linked List, Recursion
     */
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return head;
        }
        ListNode prev = null;
        ListNode cur = head;

        while (cur != null) {
            ListNode tmp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = tmp;
        }

        return prev;
    }
}
